theano/tensor/opt.py

```diff
@@ -2240,39 +2242,21 @@ def upcast(type1, type2):
     type2 from type1 constitutes an upcast.
 
     """
-    upcast_pairs = (
-        ('int8', 'int16'), ('int8', 'int32'), ('int8', 'int64'),
-        ('int16', 'int32'), ('int16', 'int64'),
-        ('int32', 'int64'),
-        ('uint8', 'uint16'), ('uint8', 'uint32'), ('uint8', 'uint64'),
-        ('uint16', 'uint32'), ('uint16', 'uint64'),
-        ('uint32', 'uint64'),
-        ('float16', 'float32'), ('float16', 'float32'), ('float16', 'float64'),
-        ('float32', 'float64'),
-        ('complex64', 'complex128'),
-
-        ('uint8', 'int16'), ('uint8', 'int32'), ('uint8', 'int64'),
-        ('uint16', 'int32'), ('uint16', 'int64'),
-        ('uint32', 'int64'),
-        ('int8', 'float16'), ('int8', 'float32'), ('int8', 'float64'),
-        ('int16', 'float32'), ('int16', 'float64'),
-        ('int32', 'float64'),
-        ('uint8', 'float16'), ('uint8', 'float32'), ('uint8', 'float64'),
-        ('uint16', 'float32'), ('uint16', 'float64'),
-        ('uint32', 'float64'),
-
-        ('int8', 'complex64'), ('int16', 'complex64'),
-        ('uint8', 'complex64'), ('uint16', 'complex64'),
-        ('float32', 'complex64'),
-        ('int8', 'complex128'), ('int16', 'complex128'), ('int32', 'complex128'),
-        ('uint8', 'complex128'), ('uint16', 'complex128'), ('uint32', 'complex128'),
-        ('float32', 'complex128'), ('float64', 'complex128')
-    )
+    category = {
+        # Pair of numbers : the 'super-index' and 'sub-index'
```

Comment:
The logic seem good, but it was hard to understand. Can you add more comments like:

super-index is an dtype (uint, int, ...) number that will only allow legal upcast of the base dtype.
sub-index is an number that represent the bits used for the exact precision numbers, to be sure we upcast the precision.